File: leetcode/python/1043_partition_array_for_maximum_sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: list[int], k: int) -> int:
        n: int = len(arr)
        dp: list[int] = [0] * (n + 1)
        for i in range(1, n + 1):
            max_val: int = 0
            for j in range(1, min(k, i) + 1):
                max_val = max(max_val, arr[i - j])
                dp[i] = max(dp[i], dp[i - j] + max_val * j)
        return dp[n]

File: leetcode/rust/src/1043_partition_array_for_maximum_sum.rs
impl Solution {
    pub fn max_sum_after_partitioning(arr: Vec<i32>, k: i32) -> i32 {
        let n = arr.len();
        let mut dp = vec![0; n + 1];

        for i in 1..=n {
            let mut max_val = 0;
            for j in 1..=k as usize {
                if i >= j {
                    max_val = max_val.max(arr[i - j]);
                    dp[i] = dp[i].max(dp[i - j] + max_val * j as i32);
                }
            }

        }

        dp[n]
    }
}